c7n/resources/ec2.py

```diff
@@ -1322,6 +1322,139 @@ def process(self, instances):
+        if not len(instances):
+            self.log.warning(
+                "action:%s instances:%r are not retained by filter" % (
+                    self.__class__.__name__.lower(),
+                    instances))
+            return
+
+        client = utils.local_session(
+            self.manager.session_factory).client('ec2')
+
+        # We list all those spot request ids ...
+        sir_values = []
+        for i in instances:
+            sir_values.append(i['SpotInstanceRequestId'])
+
+        # ... and describe the corresponding spot requests ...
+        response = client.describe_spot_instance_requests(
+            Filters=[
+                {
+                    'Name': 'spot-instance-request-id',
+                    'Values': sir_values,
+                }
+            ]
+        )
+
+        # ... to find the tags for those requests.
+        sir_tags_map = {}
+        if len(response['SpotInstanceRequests']) > 0:
+            for a in response['SpotInstanceRequests']:
+                tags = a.get('Tags', None)
+                if tags is not None:
+                    sir_tags_map[a['SpotInstanceRequestId']] = tags
+
+        # Now we find the tags we can copy : either all, either those
+        # indicated with 'only_tags' parameter.
+        copy_keys = self.data.get('only_tags', [])
+        for i in instances:
+            sir_tags = sir_tags_map.get(i['SpotInstanceRequestId'], [])
+
+            copy_tags = []
+            if copy_keys:
+                for t in sir_tags:
+                    if t['Key'] in copy_keys:
+                        copy_tags.append(t)
+            else:
+                copy_tags.extend(sir_tags)
+
+            # We won't copy aws: tags, it's forbidden and will fail
+            # the request.
+            copy_tags = list(filter(lambda x: not x['Key'].startswith('aws:'), copy_tags))
+
+            existing_tags = i.get('Tags', [])
+
+            # We may overwrite tags, but if the operation changes no tag,
+            # we will not proceed.
+            existing_tags_map = dict((h['Key'], h['Value']) for h in existing_tags)
+            copy_tags_map = dict((h['Key'], h['Value']) for h in copy_tags)
+            future_tags_map = existing_tags_map.copy()
+            future_tags_map.update(copy_tags_map)
+
+            if existing_tags_map == future_tags_map:
+                copy_tags = []
+
+            if len(copy_tags) + len(existing_tags) > self.MAX_TAG_COUNT:
+                self.log.warning(
+                    "action:%s instance:%s too many tags to copy (%d > %d)" % (
+                        self.__class__.__name__.lower(),
+                        i['InstanceId'],
+                        len(copy_tags) + len(existing_tags),
+                        self.MAX_TAG_COUNT))
+                copy_tags = []
+
+            if len(copy_tags) > 0:
+                updated_instances.append(i['InstanceId'])
+                client.create_tags(
```

Comment:
i think this could be done a bit more efficiently, ie at an api call per spot request then api call per instance.